site/app/libraries/database/DatabaseQueries.php

```diff
@@ -2268,22 +2268,54 @@ public function getAverageComponentScores($g_id, $section_key, $is_team, string
         return $return;
     }
 
-    public function getAverageGraderScores($g_id, $gc_id, $section_key, $is_team) {
+    public function getAverageGraderScores($g_id, $gc_id, $section_key, $is_team, string $bad_submissions, string $null_section) {
         $u_or_t = "u";
         $users_or_teams = "users";
         $user_or_team_id = "user_id";
+        $null_section_condition = "";
+        $bad_submissions_condition = "";
+        $bad_submission_count = 0;
+        $params = [$gc_id, $g_id, $g_id, $g_id];
         if ($is_team) {
             $u_or_t = "t";
             $users_or_teams = "gradeable_teams";
             $user_or_team_id = "team_id";
         }
+        if ($null_section !== 'include') {
+            $null_section_condition = "AND {$u_or_t}.{$section_key} IS NOT NULL";
+        }
+            $this->course_db->query(
+                "SELECT COUNT(*) as cnt
+                FROM gradeable_component AS gc, {$users_or_teams} AS {$u_or_t}
+                WHERE gc.g_id=? {$null_section_condition}
+                AND EXISTS (
+                    SELECT 1
+                    FROM late_day_cache AS ldc
+                    WHERE ldc.{$user_or_team_id} = {$u_or_t}.{$user_or_team_id}
+                    AND ldc.g_id=gc.g_id
+                    AND ldc.submission_days_late > ldc.late_day_exceptions
+                    AND ldc.late_days_change = 0
+        )",
+                [$g_id]
+            );
+            $bad_submission_count = $this->course_db->row()['cnt'];
+
+        // Check if we want to include late (bad) submissions into the average
+        if ($bad_submissions !== 'include' && $bad_submission_count > 0) {
+            $bad_submissions_condition = "INNER JOIN(
+                SELECT DISTINCT ldc.{$user_or_team_id}
+                FROM late_day_cache AS ldc
+                WHERE ldc.g_id=? AND ( submission_days_late = 0 OR ldc.late_days_change != 0 
+              ) )AS ldc ON ldc.{$user_or_team_id}={$u_or_t}.{$user_or_team_id}";
+            $params[] = $g_id;
+        }
```

Comment:
Couldn't this entire block of code be added to the giant SQL query below?  We want to minimize the number of SQL queries executed if possible.  This isn't necessarily a requested change that has to be made, just something to think about if it's easy to do.